### PR TITLE
csi: Use volfile server for fetching client-volfiles

### PR DESCRIPTION
**Pull request by vatsa287**
**What this PR does / why we need it**:

- This PR makes changes in CSI necessary to make use of volfile-server feature of brick processes,
where regeneration of client volfiles are not required, instead can fetch from any one of brick_processes.
Can add multiple brick_process to be notified as backup volfile servers. Which serves the volfiles stored in 
`/var/lib/kadalu/volfiles` inside server pods.
Can be accessed using `--volfile-server` option.

- Remove storage_options from mount flow. Later in upcoming PRs create a framework for storage_options to be configured at operator component.

**Which issue(s) this PR fixes**:
Updates: #959 

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added : No user related information are expected from this PR.
- [x] Tests updated : Existing tests are enough for this PR.
- [x] Add an entry in the `CHANGELOG.md` about the changes.

Signed-off-by: Shree Vatsa N [vatsa@kadalu.tech](mailto:vatsa@kadalu.tech)

### CODE REVIEW

File: csi/volumeutils.py [hunk 1]
```diff
@@ -940,82 +941,11 @@ def generate_client_volfile(volname):
...
-                    return
-
-    logging.debug(logf(
-        "Sent SIGHUP to glusterfs process",
-        volname=volname
-    ))
-    return
-
-
 def reload_glusterfs(volume):
```
Comment (aravindavk) [inline csi/volumeutils.py]: Remove this function itself. Both native and external volfile servers handles notifying the connected clients when any option changes.

File: csi/volumeutils.py [hunk 2]
```diff
@@ -1026,16 +956,14 @@ def reload_glusterfs(volume):
...
-        if not generate_client_volfile(volname):
+        if not has_configmap_changed(volname):
             return False
-        # TODO: ideally, keep the pid in structure for easier access
-        # pid = VOL_DATA[volname]["pid"]
-        # cmd = ["kill", "-HUP", str(pid)]
-        cmd = ["ps", "--no-header", "-ww", "-o", "pid,command", "-C", "glusterfs"]
+
+        pid = VOL_DATA[volname].get("pid", 0)
+        cmd = ["kill", "-HUP", str(pid)]
```
Comment (aravindavk) [inline csi/volumeutils.py]: L959-976 can be removed

File: csi/volumeutils.py [hunk 3]
```diff
@@ -1218,7 +1148,10 @@ def mount_glusterfs(volume, mountpoint, storage_options="", is_client=False):
                 raise err
         return mountpoint
 
-    generate_client_volfile(volname)
+    # Todo:
```
Comment (aravindavk) [inline csi/volumeutils.py]: Remove this "TODO" also. Let's remove options completely from Storage class or PV create/mount flow. Let's keep only as Storage pool configuration.

Comment (aravindavk) [reply csi/volumeutils.py]: Also remove the lines with Volfile modifications. That way we can have uniformity once NFS mount also introduced. Options can only be configured for Storage pools via operator.

**Comment by leelavg** [comment]
hi @aravindavk,

- I tried reviewing this and previous volfile PR but couldn't quite understand the logic and forward looking statements which might be introduced in upcoming PRs (like NFS support)
- I'm particularly restricting myself not to get too much involved in kadalu ecosystem, as such could you pls point what (specific) other projects should be referred for all integrations coming into this project (kadalu:devel)?

thanks.

**Comment by vatsa287** [comment]
Hi! @leelavg. 
In the previous PR of server component changes, we used [kadalu volgen library](https://github.com/kadalu/volgen]), instead of local volfile generation in server component. This was to keep all the volfile generation logic used in both kadalu-manager(moana) and Kadalu Kubernetes project as one. Also now since brick process are capable of serving volfiles to clients without glusterd, and volfile changes are handled by them, ex. SIGHUP. [Gluster FS PR](https://github.com/gluster/glusterfs/pull/3745).

This PR makes use of that feature of brick process, instead of creating volfiles again in CSI layer, we used `--volfile-server <host>` to fetch volfiles from server component, stored at `/var/lib/kadalu/volfiles/volname.vol`. And will be removing storage_options/volfile_options from here and moving the logic to be handled by operator, on storage-pool input.

Recently, we made changes to support NFS mount on kadalu_storage [issue](https://github.com/kadalu/moana/issues/65), can read [here](https://vatsa.hashnode.dev/kadalu-storage-is-now-integrated-with-nfs-ganesha) for better understanding. So with that changes done, `NFS` mount will be made as default mount rather than current `FUSE` based mount which would become optional. Changes made to NFS upstream to support kadalu storage is [here](https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/547270), rationale being that volumes would be available in MAC ecosystem, would have to try if the nodeplugin issue on upgrade would be solved with this approach.

**Comment by leelavg** [comment]
> nodeplugin issue on upgrade
- yep, it should be solved with kernel client, ie, NFS

Thanks a lot for the info, will go through the links.

File: csi/volumeutils.py (hunk 1 above)
Comment (vatsa287) [reply csi/volumeutils.py]: Removed `reload_glusterfs`.
In csi/main.py. `'reload_glusterfs` is used to reconfigure mounts. Should that be removed too?
L#54 csi/main.py

File: csi/volumeutils.py (hunk 3 above)
Comment (vatsa287) [reply csi/volumeutils.py]: Removed volume_options from mount flow.

File: csi/nodeserver.py [hunk 4]
```diff
@@ -88,22 +85,14 @@ def NodePublishVolume(self, request, context):
             'type': voltype,
         }
 
-        mountpoint = mount_glusterfs(volume, mntdir, storage_options, True)
```
Comment (vatsa287) [inline csi/nodeserver.py:91]: Removed mountpoint as it was an un-used.

**Comment by aravindavk** [comment]
> hi @aravindavk,
> 
>     * I tried reviewing this and previous volfile PR but couldn't quite understand the logic and forward looking statements which might be introduced in upcoming PRs (like NFS support)
> 
>     * I'm particularly restricting myself not to get too much involved in kadalu ecosystem, as such could you pls point what (specific) other projects should be referred for all integrations coming into this project (kadalu:devel)?
> 
> 
> thanks.

Most of it already covered by @vatsa287

## Volfile management
Currently based on the Volume info, volfiles (That is config files for glusterfs process) are generated using the templates by the GlusterFS client itself. That is CSI Provisioner and CSI node plugins. Before calling the `glusterfs` command to mount the volume, it generates the Client volfile and then passes it to `glusterfs` as `-f <volfile>`. Using volfile path instead of volfile server option makes the mount static. If any changes to the Options or new bricks list the GlusterFS clients will not know automatically. So SIGHUP and other logics needed in client side to detect that volume option changed or new bricks added etc.

In Gluster standalone setup, Glusterd serves the Volfiles to the connected clients and also keeps the connected clients list so that Glusterd can notify all the connnected clients(mounts) to reload the Volfile when any option changes. Recently Amar added Volfile server functionality to the brick processes itself. So that Gluster mounts can work as is even Glusterd doesn't exists. Brick process itself can serve Volfile via RPC to the connected clients.  This is very great feature, following things are now possible without Glusterd

- Volfile generation only server side
- Regular `mount -t glusterfs` works with Kadalu
- Brick processes keeps the connected clients list so that Volfile reload notification is automatic.
- NFS and Samba integrations will work since both are dependent on Volfile server.

New Volgen library is implemented so that it can be used by both Moana and this project. All the Volfile generation logic is handled in one place and can be used by multiple projects.

With the above two changes(volfile server and volgen library), only server pods need to generate the volfiles. CSI code becomes lighter because Volfile generation is not required and both external Gluster and native Kadalu use the same mount command.

## Storage Options:
Earlier efforts was to support Storage options via storage class, and each Gluster mount in CSI node plugin can use these options and customize the volfiles accordingly. Implementing this will make the mounts static again without any auto notification. We can support only mount options only if the access is done via fuse.

New plan is to implement Options support only for Pools (May be in Operator). Storage class may get support for mount options. I will write about this in detail.

## Access protocols
Along with the Volfile server feature, Vatsa enhanced Gluster FSAL (NFS Ganesha plugin) to understand multiple Volfile servers and different ports than `24007`. With these features, we can now add support for NFS or Samba along with fuse.

Making NFS default is not yet sure, and we need not do it. Storage Class can decide which access protocols to use.

**Comment by aravindavk** [comment]
@vatsa287 add default options here while generating the client volfile. https://github.com/kadalu/kadalu/blob/devel/server/serverutils.py#L88

```python
options = copy.copy(DEFAULT_OPTIONS)
options.update(client_data.options)

kadalu_volgen.generate(
        "/var/lib/kadalu/templates/client.vol.j2",
        data=client_data,
        options=options,
        output_file=client_volfile_path
)
```

```python
DEFAULT_OPTIONS = {
    "performance.client-io-threads": "off",
    "performance.stat-prefetch": "off",
    "performance.quick-read": "off",
    "performance.open-behind": "off",
    "performance.read-ahead": "off",
    "performance.io-cache": "off",
    "performance.readdir-ahead": "off"
}
```

**Review by aravindavk — APPROVED** [review]

**Pull request merged**